utils/container-runtime/src/container.rs

```diff
@@ -0,0 +1,443 @@
+        self.child.stdout.take()
+    }
+
+    pub fn take_stderr(&mut self) -> Option<ChildStderr> {
+        self.child.stderr.take()
+    }
+
+    pub fn exec_cmd<I, S>(&self, exec_args: ExecArgs, cmd_args: I) -> Command
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<std::ffi::OsStr>,
+    {
+        self.runtime
+            .exec_cmd(exec_args, &self.container_name, cmd_args)
+    }
+
+    pub fn exec_shell_cmd(&self, exec_args: ExecArgs, shell_cmd: impl AsRef<str>) -> Command {
+        self.runtime
+            .exec_shell_cmd(exec_args, &self.container_name, shell_cmd)
+    }
+
+    pub async fn wait_for_container(&self, timeout: Duration) -> Result<(), ContainerRuntimeError> {
+        self.runtime
+            .wait_for_container(&self.container_name, timeout)
+            .await
+    }
+
+    pub async fn try_get_host_port(
+        &self,
+        container_port: u16,
+    ) -> Result<Option<u16>, ContainerRuntimeError> {
+        self.runtime
+            .try_get_host_port(&self.container_name, container_port)
+            .await
+    }
+
+    pub async fn wait_for_host_port(
+        &self,
+        container_port: u16,
+        timeout: Duration,
+    ) -> Result<u16, ContainerRuntimeError> {
+        self.runtime
+            .wait_for_host_port(&self.container_name, container_port, timeout)
+            .await
+    }
+
+    /// Waits for specified container port to be mapped to host and start
+    /// accepting TCP connections. Returns the corresponding port on the
+    /// host network.
+    pub async fn wait_for_host_socket(
+        &self,
+        container_port: u16,
+        timeout: Duration,
+    ) -> Result<u16, ContainerRuntimeError> {
+        let start = std::time::Instant::now();
+
+        let host_port = self.wait_for_host_port(container_port, timeout).await?;
+
+        let elapsed = std::time::Instant::now() - start;
+        if elapsed >= timeout {
+            return Err(TimeoutError::new(timeout).into());
+        }
+
+        self.runtime
+            .wait_for_socket(host_port, timeout - elapsed)
+            .await?;
+
+        Ok(host_port)
+    }
+
+    pub async fn wait(&mut self) -> std::io::Result<ExitStatus> {
+        self.child.wait().await
+    }
+
+    #[tracing::instrument(level = "debug", name = "terminate_container", skip_all, fields(container_name = %self.container_name))]
+    pub async fn terminate(&mut self) -> std::io::Result<TerminateStatus> {
+        self.child.terminate(self.terminate_timeout).await
+    }
+}
+
+impl Drop for ContainerProcess {
+    fn drop(&mut self) {
+        // id is present only when process have not yet been awaited for
+        if let Some(_) = self.child.id() {
+            tracing::warn!(
+                container_name = %self.container_name,
+                "Container was not terminated - cleaning up synchronously"
+            );
+
+            let _ = self.child.terminate_blocking(self.terminate_timeout);
+        }
+    }
+}
```

Comment:
One of key places in PR. We are hoping that use will call `container.terminate().await?` for graceful shutdown, but if they forget to do so we will still clean up resources but in a blocking way.

This may stall the event loop in some circumstances, but will prevent leaking resources.

Hopefully this limitation will be removed once [AsyncDrop](https://rust-lang.github.io/async-fundamentals-initiative/roadmap/async_drop.html) problem is addressed in the language.